### PR TITLE
Enable GPU Support for Mariner AKS 2.0

### PR DESCRIPTION
**Pull request by henryli001**
**What type of PR is this?**

**What this PR does / why we need it**:

Adds GPU support for Mariner

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh [hunk 1]
```diff
@@ -25,9 +25,41 @@ installDeps() {
...
+name=CBL-Mariner Official Nvidia 2.0 x86_64
+baseurl=https://packages.microsoft.com/cbl-mariner/2.0/prod/nvidia/x86_64
+gpgkey=file:///etc/pki/rpm-gpg/MICROSOFT-RPM-GPG-KEY file:///etc/pki/rpm-gpg/MICROSOFT-METADATA-GPG-KEY
+gpgcheck=1
+repo_gpgcheck=1
+enabled=1
+skip_if_unavailable=True
+sslverify=1
+EOF
+    set -x
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:42]: leftover debug? or deliberate?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh [hunk 2]
```diff
@@ -25,9 +25,41 @@ installDeps() {
...
+
+installNvidiaContainerRuntime() {
+    for nvidia_package in nvidia-container-runtime nvidia-container-toolkit libnvidia-container-tools libnvidia-container1; do
+      if ! dnf_install 30 1 600 $nvidia_package; then
+        exit $ERR_APT_INSTALL_TIMEOUT
+      fi
+    done
+}
+
+installNvidiaDocker() {
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:59]: I don't think you need nvidia docker, mariner never supported docker and this shouldn't be required for containerd

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh [hunk 3]
```diff
@@ -25,9 +25,41 @@ installDeps() {
...
 downloadGPUDrivers() {
-    echo "GPU drivers not yet supported for Mariner"
-    exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
+    if ! dnf_install 30 1 600 cuda; then
+      exit $ERR_APT_INSTALL_TIMEOUT
+    fi
+}
+
+installNvidiaContainerRuntime() {
+    for nvidia_package in nvidia-container-runtime nvidia-container-toolkit libnvidia-container-tools libnvidia-container1; do
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:52]: this looks to be installing 1.9/3.9. this package set is correct for that version, but 3.11/1.11 added nvidia-container-toolkit-base and 1.10+ is a prereq for cgroupv2 which we should explore for mariner. 

can you possibly bump this to 3.11 to match?

```
  libnvidia-container-tools-1.9.0-3.cm2.x86_64                                  
  libnvidia-container1-1.9.0-3.cm2.x86_64                                       
  nvidia-container-runtime-3.9.0-1.cm2.x86_64                                   
  nvidia-container-toolkit-1.9.0-3.cm2.x86_64   
```

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh [hunk 4]
```diff
@@ -25,9 +25,41 @@ installDeps() {
     fi
 }
 
+addMarinerNvidiaRepo() {
+    MARINER_NVIDIA_REPO_FILEPATH="/etc/yum.repos.d/mariner-nvidia.repo"
+    touch "${MARINER_NVIDIA_REPO_FILEPATH}"
+    cat << EOF > "${MARINER_NVIDIA_REPO_FILEPATH}"
+[mariner-official-nvidia]
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:32]: any reason not to add this in vhd?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 3 above)
Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:52]: tangentially, from logs looks like this takes 60 sec on vanilla marketplace image which isn't too shabby. would be nice if we can pre-cache those via the equivalent of `apt-get install --download-only` or something to compare the times. not super critical

**Comment by alexeldeib** [comment]
mostly minor comments, thanks for the PR!

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 3 above)
Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:52]: Sure, I can bump the package versions to 1.11/3.11

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 1 above)
Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:42]: This is leftover debug. Will remove this line

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 4 above)
Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:32]: the cuda rpm is currently not supported in the Mariner core repository. Let me confirm with others in the team to make sure whether to pre-install this repo file in the Mariner image.

Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:32]: for clarity I just mean in the AKS build VHD in install-dependencies.sh, not the base marketplace image

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 3 above)
Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:51]: thanks! I'd also suggest pinning this in the script so we can control upgrades and not pull latest from mariner distro. 

unless you're planning to control it that way deliberately, but that would be a little funky imo

Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:51]: I have opened a PR to bump the related packages: https://github.com/microsoft/CBL-Mariner/pull/3813

Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Have added version enforcement when installing these packages

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 4 above)
Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Have moved this function to the VHD phase

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh [hunk 5]
```diff
@@ -26,8 +26,20 @@ installDeps() {
...
+    if ! dnf_install 30 1 600 cuda; then
+      exit $ERR_APT_INSTALL_TIMEOUT
+    fi
+}
+
+installNvidiaContainerRuntime() {
+    MARINER_NVIDIA_CONTAINER_RUNTIME_VERSION="3.11.0"
+    MARINER_NVIDIA_CONTAINER_TOOLKIT_VERSION="1.11.0"
+    
+    for nvidia_package in nvidia-container-runtime-${MARINER_NVIDIA_CONTAINER_RUNTIME_VERSION} nvidia-container-toolkit-${MARINER_NVIDIA_CONTAINER_TOOLKIT_VERSION} libnvidia-container-tools-${MARINER_NVIDIA_CONTAINER_TOOLKIT_VERSION} libnvidia-container1-${MARINER_NVIDIA_CONTAINER_TOOLKIT_VERSION}; do
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:38]: sanity check: these packages will not touch /etc/containerd/config.toml?

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh [hunk 6]
```diff
@@ -12,8 +12,22 @@ installBcc() {
     dnf_install 120 5 25 bcc-tools || exit $ERR_BCC_INSTALL_TIMEOUT
 }
 
-configGPUDrivers() {
-    echo "Not installing GPU drivers on Mariner"
+addMarinerNvidiaRepo() {
+    if [[ $OS_VERSION == "2.0" ]]; then
```
Comment (alexeldeib) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:16]: nit: a bit funny to do this here vs at the function call site in install-deps.sh

**Review by alexeldeib — APPROVED** [review]
lgtm, hopefully @hbeberman has a chance to take a look as well?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh (hunk 5 above)
Comment (henryli001) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:38]: That's correct.

**Review by hbeberman — APPROVED** [review]

**Comment by hbeberman** [comment]
:shipit:

**Pull request merged**